// @tanstack/devtools-vite/dist/esm/enhance-logs.js
import chalk from "chalk";
import { normalizePath } from "vite";
import { gen, trav } from "./babel.js";
import { parse } from "@babel/parser";
import * as t from "@babel/types";
const transform = (ast, filePath, port) => {
  let didTransform = false;
  trav(ast, {
    CallExpression(path) {
      const callee = path.node.callee;
      if (callee.type === "MemberExpression" && callee.object.type === "Identifier" && callee.object.name === "console" && callee.property.type === "Identifier" && (callee.property.name === "log" || callee.property.name === "error")) {
        const location = path.node.loc;
        if (!location) {
          return;
        }
        const [lineNumber, column] = [
          location.start.line,
          location.start.column
        ];
        const finalPath = `${filePath}:${lineNumber}:${column + 1}`;
        const logMessage = `${chalk.magenta("LOG")} ${chalk.blueBright(`${finalPath}`)}
 → `;
        const serverLogMessage = t.arrayExpression([
          t.stringLiteral(logMessage)
        ]);
        const browserLogMessage = t.arrayExpression([
          // LOG with css formatting specifiers: %c
          t.stringLiteral(
            `%c${"LOG"}%c %c${`Go to Source: http://localhost:${port}/__tsd/open-source?source=${encodeURIComponent(finalPath)}`}%c 
 → `
          ),
          // magenta
          t.stringLiteral("color:#A0A"),
          t.stringLiteral("color:#FFF"),
          // blueBright
          t.stringLiteral("color:#55F"),
          t.stringLiteral("color:#FFF")
        ]);
        const checkServerCondition = t.binaryExpression(
          "===",
          t.unaryExpression("typeof", t.identifier("window")),
          t.stringLiteral("undefined")
        );
        path.node.arguments.unshift(
          t.spreadElement(
            t.conditionalExpression(
              checkServerCondition,
              serverLogMessage,
              browserLogMessage
            )
          )
        );
        didTransform = true;
      }
    }
  });
  return didTransform;
};
function enhanceConsoleLog(code, id, port) {
  const [filePath] = id.split("?");
  const location = filePath?.replace(normalizePath(process.cwd()), "");
  try {
    const ast = parse(code, {
      sourceType: "module",
      plugins: ["jsx", "typescript"]
    });
    const didTransform = transform(ast, location, port);
    if (!didTransform) {
      return;
    }
    return gen(ast, {
      sourceMaps: true,
      retainLines: true,
      filename: id,
      sourceFileName: filePath
    });
  } catch (e) {
    return;
  }
}
export {
  enhanceConsoleLog
};
