// @tanstack/db/dist/esm/index.js
import * as ir from "./query/ir.js";
import { CollectionImpl, createCollection } from "./collection/index.js";
import { SortedMap } from "./SortedMap.js";
import { createTransaction, getActiveTransaction } from "./transactions.js";
import { createArrayChangeProxy, createChangeProxy, withArrayChangeTracking, withChangeTracking } from "./proxy.js";
import { createOptimisticAction } from "./optimistic-action.js";
import { localOnlyCollectionOptions } from "./local-only.js";
import { localStorageCollectionOptions } from "./local-storage.js";
import { AggregateFunctionNotInSelectError, AggregateNotSupportedError, CannotCombineEmptyExpressionListError, CollectionConfigurationError, CollectionInErrorStateError, CollectionInputNotFoundError, CollectionIsInErrorStateError, CollectionOperationError, CollectionRequiresConfigError, CollectionRequiresSyncConfigError, CollectionStateError, DeleteKeyNotFoundError, DistinctRequiresSelectError, DuplicateAliasInSubqueryError, DuplicateDbInstanceError, DuplicateKeyError, DuplicateKeySyncError, EmptyReferencePathError, GroupByError, HavingRequiresGroupByError, InvalidCollectionStatusTransitionError, InvalidJoinCondition, InvalidJoinConditionLeftSourceError, InvalidJoinConditionRightSourceError, InvalidJoinConditionSameSourceError, InvalidJoinConditionSourceMismatchError, InvalidKeyError, InvalidSchemaError, InvalidSourceError, InvalidSourceTypeError, InvalidStorageDataFormatError, InvalidStorageObjectFormatError, InvalidWhereExpressionError, JoinCollectionNotFoundError, JoinConditionMustBeEqualityError, JoinError, KeyUpdateNotAllowedError, LimitOffsetRequireOrderByError, LocalStorageCollectionError, MissingAliasInputsError, MissingDeleteHandlerError, MissingHandlerError, MissingInsertHandlerError, MissingMutationFunctionError, MissingUpdateArgumentError, MissingUpdateHandlerError, NegativeActiveSubscribersError, NoKeysPassedToDeleteError, NoKeysPassedToUpdateError, NoPendingSyncTransactionCommitError, NoPendingSyncTransactionWriteError, NonAggregateExpressionNotInGroupByError, NonRetriableError, OnMutateMustBeSynchronousError, OnlyOneSourceAllowedError, QueryBuilderError, QueryCompilationError, QueryMustHaveFromClauseError, QueryOptimizerError, SchemaMustBeSynchronousError, SchemaValidationError, SerializationError, SetWindowRequiresOrderByError, StorageError, StorageKeyRequiredError, SubQueryMustHaveFromClauseError, SubscriptionNotFoundError, SyncCleanupError, SyncTransactionAlreadyCommittedError, SyncTransactionAlreadyCommittedWriteError, TanStackDBError, TransactionAlreadyCompletedRollbackError, TransactionError, TransactionNotPendingCommitError, TransactionNotPendingMutateError, UndefinedKeyError, UnknownExpressionTypeError, UnknownFunctionError, UnknownHavingExpressionTypeError, UnsupportedAggregateFunctionError, UnsupportedFromTypeError, UnsupportedJoinSourceTypeError, UnsupportedJoinTypeError, UpdateKeyNotFoundError, WhereClauseConversionError } from "./errors.js";
import { deepEquals } from "./utils.js";
import { createPacedMutations } from "./paced-mutations.js";
import { BaseIndex, IndexOperation } from "./indexes/base-index.js";
import { BTreeIndex } from "./indexes/btree-index.js";
import { IndexProxy, LazyIndexWrapper } from "./indexes/lazy-index.js";
import { extractFieldPath, extractSimpleComparisons, extractValue, parseLoadSubsetOptions, parseOrderByExpression, parseWhereExpression, walkExpression } from "./query/expression-helpers.js";
import { add, and, avg, coalesce, concat, count, eq, gt, gte, ilike, inArray, isNull, isUndefined, length, like, lower, lt, lte, max, min, not, operators, or, sum, upper } from "./query/builder/functions.js";
import { BaseQueryBuilder, Query } from "./query/builder/index.js";
import { compileQuery } from "./query/compiler/index.js";
import { createLiveQueryCollection, liveQueryCollectionOptions } from "./query/live-query-collection.js";
import { isLimitSubset, isOffsetLimitSubset, isOrderBySubset, isPredicateSubset, isWhereSubset, minusWherePredicates, unionWherePredicates } from "./query/predicate-utils.js";
import { DeduplicatedLoadSubset } from "./query/subset-dedupe.js";
import { debounceStrategy } from "./strategies/debounceStrategy.js";
import { queueStrategy } from "./strategies/queueStrategy.js";
import { throttleStrategy } from "./strategies/throttleStrategy.js";
export {
  AggregateFunctionNotInSelectError,
  AggregateNotSupportedError,
  BTreeIndex,
  BaseIndex,
  BaseQueryBuilder,
  CannotCombineEmptyExpressionListError,
  CollectionConfigurationError,
  CollectionImpl,
  CollectionInErrorStateError,
  CollectionInputNotFoundError,
  CollectionIsInErrorStateError,
  CollectionOperationError,
  CollectionRequiresConfigError,
  CollectionRequiresSyncConfigError,
  CollectionStateError,
  DeduplicatedLoadSubset,
  DeleteKeyNotFoundError,
  DistinctRequiresSelectError,
  DuplicateAliasInSubqueryError,
  DuplicateDbInstanceError,
  DuplicateKeyError,
  DuplicateKeySyncError,
  EmptyReferencePathError,
  GroupByError,
  HavingRequiresGroupByError,
  ir as IR,
  IndexOperation,
  IndexProxy,
  InvalidCollectionStatusTransitionError,
  InvalidJoinCondition,
  InvalidJoinConditionLeftSourceError,
  InvalidJoinConditionRightSourceError,
  InvalidJoinConditionSameSourceError,
  InvalidJoinConditionSourceMismatchError,
  InvalidKeyError,
  InvalidSchemaError,
  InvalidSourceError,
  InvalidSourceTypeError,
  InvalidStorageDataFormatError,
  InvalidStorageObjectFormatError,
  InvalidWhereExpressionError,
  JoinCollectionNotFoundError,
  JoinConditionMustBeEqualityError,
  JoinError,
  KeyUpdateNotAllowedError,
  LazyIndexWrapper,
  LimitOffsetRequireOrderByError,
  LocalStorageCollectionError,
  MissingAliasInputsError,
  MissingDeleteHandlerError,
  MissingHandlerError,
  MissingInsertHandlerError,
  MissingMutationFunctionError,
  MissingUpdateArgumentError,
  MissingUpdateHandlerError,
  NegativeActiveSubscribersError,
  NoKeysPassedToDeleteError,
  NoKeysPassedToUpdateError,
  NoPendingSyncTransactionCommitError,
  NoPendingSyncTransactionWriteError,
  NonAggregateExpressionNotInGroupByError,
  NonRetriableError,
  OnMutateMustBeSynchronousError,
  OnlyOneSourceAllowedError,
  Query,
  QueryBuilderError,
  QueryCompilationError,
  QueryMustHaveFromClauseError,
  QueryOptimizerError,
  SchemaMustBeSynchronousError,
  SchemaValidationError,
  SerializationError,
  SetWindowRequiresOrderByError,
  SortedMap,
  StorageError,
  StorageKeyRequiredError,
  SubQueryMustHaveFromClauseError,
  SubscriptionNotFoundError,
  SyncCleanupError,
  SyncTransactionAlreadyCommittedError,
  SyncTransactionAlreadyCommittedWriteError,
  TanStackDBError,
  TransactionAlreadyCompletedRollbackError,
  TransactionError,
  TransactionNotPendingCommitError,
  TransactionNotPendingMutateError,
  UndefinedKeyError,
  UnknownExpressionTypeError,
  UnknownFunctionError,
  UnknownHavingExpressionTypeError,
  UnsupportedAggregateFunctionError,
  UnsupportedFromTypeError,
  UnsupportedJoinSourceTypeError,
  UnsupportedJoinTypeError,
  UpdateKeyNotFoundError,
  WhereClauseConversionError,
  add,
  and,
  avg,
  coalesce,
  compileQuery,
  concat,
  count,
  createArrayChangeProxy,
  createChangeProxy,
  createCollection,
  createLiveQueryCollection,
  createOptimisticAction,
  createPacedMutations,
  createTransaction,
  debounceStrategy,
  deepEquals,
  eq,
  extractFieldPath,
  extractSimpleComparisons,
  extractValue,
  getActiveTransaction,
  gt,
  gte,
  ilike,
  inArray,
  isLimitSubset,
  isNull,
  isOffsetLimitSubset,
  isOrderBySubset,
  isPredicateSubset,
  isUndefined,
  isWhereSubset,
  length,
  like,
  liveQueryCollectionOptions,
  localOnlyCollectionOptions,
  localStorageCollectionOptions,
  lower,
  lt,
  lte,
  max,
  min,
  minusWherePredicates,
  not,
  operators,
  or,
  parseLoadSubsetOptions,
  parseOrderByExpression,
  parseWhereExpression,
  queueStrategy,
  sum,
  throttleStrategy,
  unionWherePredicates,
  upper,
  walkExpression,
  withArrayChangeTracking,
  withChangeTracking
};
